pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1459 @@
+				initial_price,
+			});
+
+			Ok(())
+		}
+
+		/// Add liquidity of asset `asset` in quantity `amount` to Omnipool
+		///
+		/// `add_liquidity` adds specified asset amount to pool and in exchange gives the origin
+		/// corresponding shares amount in form of NFT at current price.
+		///
+		/// Asset's tradable state must contain ADD_LIQUIDITY flag, otherwise `NotAllowed` error is returned.
+		///
+		/// NFT is minted using NTFHandler which implements non-fungibles traits from frame_support.
+		///
+		/// Parameters:
+		/// - `asset`: The identifier of the new asset added to the pool. Must be already in the pool
+		/// - `amount`: Amount of asset added to omnipool
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(origin: OriginFor<T>, asset: T::AssetId, amount: Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinimumPoolLiquidity::get(),
+				Error::<T>::InsufficientLiquidity
+			);
+
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let stable_asset = Self::stable_asset()?;
+
+			let asset_state = Self::load_asset_state(asset)?;
+
+			ensure!(
+				asset_state.tradable.contains(Tradability::ADD_LIQUIDITY),
+				Error::<T>::NotAllowed
+			);
+
+			let state_changes = hydra_dx_math::omnipool::calculate_add_liquidity_state_changes(
+				&(&asset_state).into(),
+				amount,
+				stable_asset,
+				asset == T::StableCoinAssetId::get(),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			let new_asset_state = asset_state
+				.delta_update(&state_changes.asset)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			let hub_reserve_ratio = FixedU128::checked_from_rational(
+				new_asset_state.hub_reserve,
+				T::Currency::free_balance(T::HubAssetId::get(), &Self::protocol_account())
+					.checked_add(*state_changes.asset.delta_hub_reserve)
+					.ok_or(ArithmeticError::Overflow)?,
+			)
+			.ok_or(ArithmeticError::DivisionByZero)?;
+
+			ensure!(
+				hub_reserve_ratio <= Self::asset_weight_cap(),
+				Error::<T>::AssetWeightCapExceeded
+			);
+
+			let updated_asset_price = new_asset_state.price().ok_or(ArithmeticError::DivisionByZero)?;
+
+			// Create LP position with given shares
+			let lp_position = Position::<Balance, T::AssetId> {
+				asset_id: asset,
+				amount,
+				shares: *state_changes.asset.delta_shares,
+				// Note: position needs price after asset state is updated.
+				price: updated_asset_price.into_inner(),
+			};
+
+			let instance_id = Self::create_and_mint_position_instance(&who)?;
+
+			<Positions<T>>::insert(instance_id, lp_position);
+
+			Self::deposit_event(Event::PositionCreated {
+				position_id: instance_id,
+				owner: who.clone(),
+				asset,
+				amount,
+				shares: *state_changes.asset.delta_shares,
+				price: updated_asset_price,
+			});
+
+			T::Currency::transfer(
```

Comment:
everywhere we are assuming this transfers the exact amount of tokens to the protocol account, which doesn't have to be always a case and you are not in control of the actual implementation, we should always check that actual asset balances are consistent with whatever we are tracking in asset state (we have enough available balance)

Reply:
discussed "offline". Not something omnipool should be considering. It relies on correct MultiCurrency trait implemenation provided.